staging/src/k8s.io/client-go/tools/auth/exec/exec_test.go

```diff
@@ -0,0 +1,189 @@
+					},
+				},
+			}),
+			wantExecCredential: &clientauthenticationv1beta1.ExecCredential{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "ExecCredential",
+					APIVersion: clientauthenticationv1beta1.SchemeGroupVersion.String(),
+				},
+				Spec: clientauthenticationv1beta1.ExecCredentialSpec{
+					Cluster: &clientauthenticationv1beta1.Cluster{
+						Server:                   "https://some-server/some/path",
+						TLSServerName:            "some-server-name",
+						InsecureSkipTLSVerify:    true,
+						CertificateAuthorityData: []byte("some-ca-data"),
+						ProxyURL:                 "https://some-proxy-url:12345",
+					},
+				},
+			},
+			wantRESTInfo: restInfo{
+				host: "https://some-server/some/path",
+				tlsClientConfig: rest.TLSClientConfig{
+					Insecure:   true,
+					ServerName: "some-server-name",
+					CAData:     []byte("some-ca-data"),
+				},
+				proxyURL: "https://some-proxy-url:12345",
+			},
+		},
+		{
+			name:            "v1beta1 nil cluster",
+			data:            marshal(t, clientauthenticationv1beta1.SchemeGroupVersion, &clientauthenticationv1beta1.ExecCredential{}),
+			wantErrorPrefix: "exec credential does not contain cluster information",
+		},
+		{
+			name:            "v1alpha1",
+			data:            marshal(t, clientauthenticationv1alpha1.SchemeGroupVersion, &clientauthenticationv1alpha1.ExecCredential{}),
+			wantErrorPrefix: "exec credential does not contain cluster information",
+		},
+		{
+			name:            "invalid object kind",
+			data:            marshal(t, corev1.SchemeGroupVersion, &corev1.Pod{}),
+			wantErrorPrefix: "invalid group/kind: wanted ExecCredential.client.authentication.k8s.io, got Pod",
+		},
+		{
+			name:            "bad data",
+			data:            []byte("bad data"),
+			wantErrorPrefix: "decode: ",
+		},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			execCredential, restConfig, err := LoadExecCredential(test.data)
+			if test.wantErrorPrefix != "" {
+				if err == nil {
+					t.Error("wanted error, got success")
+				} else if !strings.HasPrefix(err.Error(), test.wantErrorPrefix) {
+					t.Errorf("wanted '%s', got '%s'", test.wantErrorPrefix, err.Error())
+				}
+			} else if err != nil {
+				t.Error(err)
+			} else {
+				match(t, test.wantExecCredential, execCredential)
+
+				match(t, test.wantRESTInfo.host, restConfig.Host)
+				match(t, test.wantRESTInfo.tlsClientConfig, restConfig.TLSClientConfig)
+
+				proxyURL, err := restConfig.Proxy(nil)
+				if err != nil {
+					t.Fatal(err)
+				}
+				match(t, test.wantRESTInfo.proxyURL, proxyURL.String())
+			}
+		})
+	}
+}
+
+func marshal(t *testing.T, gv schema.GroupVersion, obj runtime.Object) []byte {
+	t.Helper()
+
+	data, err := runtime.Encode(codecs.LegacyCodec(gv), obj)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return data
+}
+
+// TODO: are there global test helpers that do this?
+func match(t *testing.T, expected, actual interface{}) {
```

Comment:
Are there well-known kube testing utilities that do this? The best I could do was find that `diff` package...

Reply:
cmp.Diff(a, b) is what we normally use

Reply:
https://github.com/kubernetes/kubernetes/pull/95489/commits/bd4af0df52b7da2a838795ccf9a8760a0316da48